### DISCUSSION TITLE
Terrain editor

### DISCUSSION THREAD

**Discussion by larinius** [stride3d/stride#1330] [Ideas]
Hire my short list of features for "Terrain Editor"

1)	Input format – png/raw heightmaps as it allows use existing maps, import maps from Geo websites  (most services export heightmaps), use world editors  (Worldmashine, Worldgenerator etc). Internally, the engine can convert it into more convenient format for further processing.

2)	Automatically split big heightmaps (4000-8000 px) into smaller tiles (517-1025px).
a)	Dynamically load the nearest terrain tiles based on Camera position. 
(This will help to create big sized levels without LOD)

3)	Brush for material painting with set of most usable shapes.
a)	Option to set specific height and angle for material layers 
(“Auto materials”, when dirt/gras/rocks automatically applied based on terrain shape).

4)	Brush for foliage painting with randomization presets (scattered evenly/random/ in groups).
a)	“Foliage palettes” – save several assets into a palette and use it with brush later. 
(For example, for painting different biomes).

5)	Brush for heightmap editing (rise, lower, flatten, erosion, smooth).
a)	Option to choose brush shapes from presets.

**Comment by tebjan** [comment]
also for reference: https://github.com/stride3d/stride/issues/54

**Comment by MechWarrior99** [comment]
Here are my thoughts and requests for a terrain system.

Being able to use a height map is something I agree is important, they are easy to work with and allow for using other software to generate/design the terrain.
However a simple height map based tessellation terrain system has some drawbacks. They don't support caves well, they don't support overhangs, and can lack the resolution required for more detailed terrain.

The other obvious option is voxel terrain, however it generally has worse performance, is harder to edit, and also lacks the ability to show high levels of detail.

I believe investing time in researching alternative methods and techniques would be well worth it, to be able to have a better terrain system.
It would also be good to see what other engines do and what users like or dislike about them.

To me ensuring that the terrain is performant and scalable for all times of games, from small puzzle games to large open worlds is highly important. If it is not then later on we will simply need to create another one that is and that will result in wasted effort and time.

Vegetation was mentioned. And while I think that is very important. I also feel it is a some what separate issue from terrain but should be kept in mind still.

I also know I personally would love a somewhat non-destructive workflow for terrain editing.

These are my rambling input on the matter of terrain.

**Comment by shinkitsunemaru** [comment]
We have some of those features mentioned on https://github.com/johang88/TR.Stride/tree/master/TR.Stride.Terrain

Kudos to @johang88
